Global work package create

oliverguenther opened the pull request:
PR [#23067](https://community.openproject.com/work_packages/23067/activity)

Use the global create form for the create form (inplace and inline).
- Adds the `project` attribute to the inplace group
- Retrieves the apiV3 project path from the projectIdentifier (through the experimental API 👎 ) since I don't have the V3 project id available.
- **Removes the type dropdown**, since we don't have types in the global create.
- Adds a global `/create_new` route to the split-screen create

This PR also changes the authorization check from checking availability of available_projects to the `createWorkPackage` link in the wp collection response.

**Issue**

Despite `type` not being shown in the create form by default, since it is marked as `hasDefault`. We do not have a default value when initially creating the work package, so type should not be marked as such.

Still, the type has a default value `-` since it is marked null. This will be resolved when switching to the new inline-edit fields, so I don't think it makes since fixing it now.

![demo_project_-_work_packages_-_openproject](https://cloud.githubusercontent.com/assets/459462/15072440/08945462-1393-11e6-8576-53fa5d0072a2.png)

oliverguenther committed:
Use create link in wp collection to decide authorization

oliverguenther committed:
Remove type dropdown

oliverguenther committed:
Add project as editable field

oliverguenther committed:
Always post to global create form

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-inline-create-global-requests.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-inline-create-global-requests.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

oliverguenther committed:
Do not mark type as hasDefault

oliverguenther committed:
Add global /create_new route

machisuji commented:
@oliverguenther yes that's definitely and error then. I think it's just hard-coded in the API since I thought it always would have a value. Seeing as it does when one [creates](https://community.openproject.com/projects/gmbh/work_packages/create_new) a new work package. Where is this global view then and why does it not have one there?

oliverguenther commented:
This PR introduces the global create view, which does not have a type available for the initial form post, and thus the API responds with an empty value. I've adjusted the type attribute accordingly.

machisuji committed:
specs have to set the type now

machisuji committed:
fixed routing

pagination was broken due to this
links to work packages were generated with /create_new in them

machisuji committed:
use bundler to run cucumber

oliverguenther committed:
recreated broken and incomplete cukes as specs

oliverguenther committed:
Correct allowed state of creat button based on links

We can derive the :add_work_package permission as follows:

1. List: work_packages collection has a `createWorkPackage` link when
allowed.
2. Single Work package: A single work package has a `addChild` that is
available when add work packages is available.

oliverguenther committed:
Fix global create spec

oliverguenther committed:
Create mode: Fix switch to full view

oliverguenther committed:
Fix new spec

oliverguenther committed:
Ugly hack for fixing switching types

Switching types may reload the fields while selenium enters data into
it.

oliverguenther committed:
Remove invalid description spec

oliverguenther committed:
Fix focussing on erroneous fields

oliverguenther committed:
Fix description focussing
